src/modules/0.0.15/aws_elasticache/index.ts

```diff
@@ -0,0 +1,275 @@
+                NumCacheNodes: cluster.numNodes,
+              };
+              const res: CacheClusterAWS | undefined = await createCacheCluster(
+                client.elasticacheClient,
+                input
+              );
+              if (res) {
+                const newCluster: CacheCluster =
+                  await AwsElastiCacheModule.utils.cacheClusterMapper(res, ctx);
+                newCluster.clusterId = cluster.clusterId;
+                await AwsElastiCacheModule.mappers.cacheCluster.db.update(
+                  newCluster,
+                  ctx
+                );
+                out.push(newCluster);
+              }
+            }
+            return out;
+          },
+          read: async (ctx: Context, clusterId?: string) => {
+            const client = (await ctx.getAwsClient()) as AWS;
+            if (clusterId) {
+              const rawCluster = await getCacheCluster(
+                client.elasticacheClient,
+                clusterId
+              );
+              if (rawCluster?.CacheClusterStatus === "deleting")
+                return undefined;
+
+              return AwsElastiCacheModule.utils.cacheClusterMapper(
+                rawCluster,
+                ctx
+              );
+            } else {
+              const rawClusters =
+                (await getCacheClusters(client.elasticacheClient)) ?? [];
+              const out = [];
+              for (const i of rawClusters) {
+                if (i.CacheClusterStatus === "deleting") continue;
+                out.push(
+                  await AwsElastiCacheModule.utils.cacheClusterMapper(i, ctx)
+                );
+              }
+              return out;
+            }
+          },
+          update: async (clusters: CacheCluster[], ctx: Context) => {
+            // if user has modified state, restore it. If not, go with replace path
+            const client = (await ctx.getAwsClient()) as AWS;
+            const out = [];
+            for (const cluster of clusters) {
+              const cloudRecord =
+                ctx?.memo?.cloud?.CacheCluster?.[cluster.clusterId ?? ""];
+              const isUpdate = Object.is(
+                AwsElastiCacheModule.mappers.cacheCluster.cloud.updateOrReplace(
+                  cloudRecord,
+                  cluster
+                ),
+                "update"
+              );
+              if (!isUpdate) {
+                // we cannot modify the engine, restore
+                if (cluster.engine !== cloudRecord.engine) {
+                  cluster.engine = cloudRecord.engine;
+                  await AwsElastiCacheModule.mappers.cacheCluster.db.update(
+                    cluster,
+                    ctx
+                  );
+                  out.push(cluster);
+                } else {
+                  // if the cluster ID is different, no needed to delete it
+                  if (cluster.clusterId===cloudRecord.clusterId) {
```

Comment:
just as a comment, this check is not necessary, it should never come through the update path since it will be detected as create/delete in the main apply path.

Reply:
mm from my understanding i thought it will fall under update, because equals compares node type, num nodes and engine. But actually i'm thinking that for this case, equals needs to be based on clusterId directly... i can have multiple clusters with same engine, nodes ... and it doesn't mean that is the same. What do you think?

Reply:
`equals` determines if two things that are identified as the same thing are actually equal or not. `@cloudId` (usually) is what is used to decide what things should be compared and have CRUD operations potentially performed on them.

Reply:
So the properties you are comparing in the `mapper.equals` method are the ones that will trigger any `mapper.update` call if there's any difference between objects with the same id. In this case, `clusterId` is the cloudId, meaning is what we use to identify cloud resources as unique. When an apply occurs it first matches based on cloudIds in order to see what is only defined in AWS (then need to be deleted), what is only in the DB (then it needs to be created) and for objects with the same cloudId call the `mappers.equals`, see if they have differences and then apply the `mapper.update` if necessary.